deepspeed/module_inject/replace_module.py

```diff
@@ -398,16 +399,104 @@ def _replace(child, name, conv_linear_layer):
                                          dtype=child.weight.dtype)
                 if conv_linear_layer:
                     child.weight.data = child.weight.data.transpose(-1, -2).contiguous()
-                data = mp_replace.copy(new_weight, child.weight.data)
 
                 new_bias = torch.empty((weight_shape[0] // mp_size),
                                        device=child.weight.device,
                                        dtype=child.weight.dtype)
-                bias_data = None if child.bias is None else mp_replace.copy(new_bias, child.bias.data).to(
-                    get_accelerator().current_device_name())
+
+                if require_tp_fused_qkvw(name):
+                    #for detecting fused type
+                    module_str = str(module).strip()
+                    data = mp_replace.copy(new_weight, prepare_tp_fused_qkvw(module_str, child.weight.data, mp_size))
+                    bias_data = None if child.bias is None else mp_replace.copy(
+                        new_bias, prepare_tp_fused_qkvw(module_str, child.bias.data, mp_size)).to(
+                            get_accelerator().current_device_name())
+                else:
+                    data = mp_replace.copy(new_weight, child.weight.data)
+                    bias_data = None if child.bias is None else mp_replace.copy(new_bias, child.bias.data).to(
+                        get_accelerator().current_device_name())
+
                 setattr(child, "replaced", True)
                 return LinearLayer(weight=data.to(get_accelerator().current_device_name()), bias=bias_data)
 
+        def require_tp_fused_qkvw(name):
```

Comment:
@inkcherry, can we please move all these logic into a separate file, so that we can then refactor the tp-sharding logic based on this? thanks